File: CSharp/LeetCode/Solutions/2530-maximal-score-after-applying-k-operations.cs
﻿using LeetCode.Solutions.Shared;

namespace LeetCode.Solutions;

internal class Solution2530 : IRunProgram
{
    public void Run()
    {

    }

    public long MaxKelements(int[] nums, int k)
    {
        PriorityQueue<int, int> queue = new PriorityQueue<int, int>(nums.Select(x => (x, x)), Comparer<int>.Create((a, b) => b - a));
        long output = 0L;

        for (int i = 0; i < k; i++)
        {
            int value = queue.Dequeue();
            output += value;
            value = (int)Math.Ceiling(value * 1.0 / 3);
            queue.Enqueue(value, value);
        }

        return output;
    }
}

File: Java/2530-maximal-score-after-applying-k-operations.java
class Solution {
    public long maxKelements(int[] nums, int k) {
        PriorityQueue<Integer> queue = new PriorityQueue<>();
        for (int num : nums) {
            queue.add(-num);
        }

        long output = 0L;
        for (; k > 0; k--) {
            int val = queue.poll();
            output += -val;
            queue.add((int)Math.floor(val * 1.0 / 3));
        }

        return output;
    }
}